Vagrantfile

```diff
@@ -1,71 +1,320 @@
+  define_ubuntu config, name: 'bionic64', dist: 'bionic', bits: 64
+  define_ubuntu config, name: 'xenial64', dist: 'xenial', bits: 64
+  define_ubuntu config, name: 'xenial32', dist: 'xenial', bits: 32
+  # llvm packages > 3.8 are not available for precise/trusty. Using pre built packages
+  define_ubuntu config, name: 'trusty64', dist: 'trusty', bits: 64, llvm: { url: "http://crystal-lang.s3.amazonaws.com/llvm/llvm-3.9.1-1-linux-x86_64.tar.gz", path: "llvm-3.9.1-1" }
+  define_ubuntu config, name: 'trusty32', dist: 'trusty', bits: 32, llvm: { url: "http://crystal-lang.s3.amazonaws.com/llvm/llvm-3.9.1-1-linux-i686.tar.gz", path: "llvm-3.9.1-1" }
+  define_ubuntu config, name: 'precise64', dist: 'precise', bits: 64, llvm: { url: "http://crystal-lang.s3.amazonaws.com/llvm/llvm-3.9.1-1-linux-x86_64.tar.gz", path: "llvm-3.9.1-1" }
+  define_ubuntu config, name: 'precise32', dist: 'precise', bits: 32, llvm: { url: "http://crystal-lang.s3.amazonaws.com/llvm/llvm-3.9.1-1-linux-i686.tar.gz", path: "llvm-3.9.1-1" }
+
+  define_debian config, name: 'stretch64', dist: 'stretch', bits: 64
+  define_debian config, name: 'jessie64', dist: 'jessie', bits: 64
+
+  define_freebsd config, name: 'freebsd11', box: '11.2-STABLE'
+
+  define_alpine config, name: 'alpine64', bits: 64
+
+  config.vm.provider "virtualbox" do |vb|
+    vb.cpus = ENV.fetch("VM_CPUS", "2").to_i
+
+    # Keep time synced with host
+    # vb.customize [ "guestproperty", "set", :id, "/VirtualBox/GuestAdd/VBoxService/--timesync-set-threshold", 10000 ]
+  end
+end
+
+def clone_crystal_from_vagrant(config)
+  # use ~/crystal directory instead of /vagrant
+  # to have a clean copy of working directory
+  # in the native fs of the vm
   config.vm.provision :shell, privileged: false, inline: %(
     git clone /vagrant crystal
   )
 end
 
-Vagrant.configure("2") do |config|
-  %w(precise trusty xenial).product([32, 64]).each do |dist, bits|
-    box_name = "#{dist}#{bits}"
+def setup_memory(config, bits)
+  memory = ENV.fetch("VM_MEMORY_MB", "4096").to_i
+  memory = [memory, 4096].min if bits == 32
+
+  config.vm.provider "virtualbox" do |vb|
+    vb.memory = memory
+  end
+end
 
-    config.vm.define(box_name) do |c|
-      c.vm.box = "ubuntu/#{box_name}"
+# *llvm* values: 6.0, 7, (empty), Hash with {url:, path:}
+def install_llvm(dist, llvm)
+  llvm ||= "6.0"
+
+  if llvm.is_a?(Hash)
+    %(
+      curl -s #{llvm[:url]} | tar xz -C /opt
+      echo 'export PATH="$PATH:/opt/#{llvm[:path]}/bin"' >> /etc/profile.d/crystal.sh
+    )
+  else
+    llvm_suffix = llvm != "" ? "-#{llvm}" : ""
+
+    %(
+      add-apt-repository "deb http://apt.llvm.org/#{dist}/ llvm-toolchain-#{dist}#{llvm_suffix} main"
+      curl -sSL https://apt.llvm.org/llvm-snapshot.gpg.key | apt-key add -
+      apt-get update
+      apt-get install -y llvm#{llvm_suffix}-dev
+    )
+  end
+end
+
+# instructions to install crystal binary on all platforms
+def install_crystal(c, family, dist:, bits:, github_targz:)
```

Comment:
This method feels very convoluted. It's hard to follow the control flow from individual define methods here and then into different branches again.

I'd prefer to have an `install_from_targz` method and integrate the familiy-specific installation instructions into the respective `define_<family>` methods, maybe separate the install step into `install_crystal_<family>`.